packages/dds/tree/src/feature-libraries/persisted-config/persistedConfig.ts

```diff
@@ -0,0 +1,336 @@
+	}
+
+	tryProcessOp(
+		message: ISequencedDocumentMessage,
+		local: boolean,
+		localOpMetadata: unknown,
+	): boolean {
+		this.processMinSeq(message);
+		if (local) {
+			this.pendingOpCount--;
+		}
+		const { contents } = message;
+		if (isUpgradeOp(contents)) {
+			if (contents.config.protocolIteration > this.config.protocolIteration) {
+				// TODO: validate proposed config is understood by this version of the code, fail if not.
+				if (this.requiresOpResubmission(contents.config, this.config)) {
+					this.mostRecentResubmissionSeq = message.sequenceNumber;
+
+					if (this.pendingOpCount > 0) {
+						// not technically necessary given event takes the argument, but less confusing for consumer--puts
+						// this class in a consistent state before calling outside.
+						this.config = contents.config;
+						// TODO: This has potential for dangerous op re-entrancy. need to have a discussion about it.
+						this.reSubmitPendingOps(contents.config);
+					}
+				}
+				this.updateConfig(contents.config);
+			} else if (contents.config.protocolIteration === this.config.protocolIteration) {
+				if (!areEquivalentConfigs(contents.config, this.config)) {
+					throw new Error("TODO: determine error policy here");
+				}
+			}
+			// otherwise, proposed protocol should be ignored. this could happen with a client submitting a concurrent upgrade
+			// request
+
+			if (
+				contents.config.protocolIteration >=
+				(this.configAwaitingSubmission?.protocolIteration ?? 0)
+			) {
+				// TODO: Could consider validating deep equality of the config in the case of protocol equality.
+				// This avoids sending duplicate upgrade ops in most cases (they can still happen due to concurrency).
+				// Since clients should process duplicate upgrade ops as no-ops (or ignore them, if upgrading to an older
+				// protocol iteration), this step is technically unnecessary. But it avoids some network traffic.
+				this.configAwaitingSubmission = undefined;
+			}
+
+			return true;
+		}
+
+		if (
+			this.mostRecentResubmissionSeq !== undefined &&
+			this.mostRecentResubmissionSeq > message.referenceSequenceNumber
+		) {
+			// Mark this op as handled: its format is no longer valid to be interpreted. The submitter of this op is
+			// expected to resubmit it (which they will do when they process the upgrade message).
+			return true;
+		}
+
+		// TODO: throw on ops which use a write version or a feature this client doesn't understand
+		// TODO: emit telemetry sufficient to diagnose issues
+		return false;
+	}
+
+	public getConfigForNextSubmission(): PersistedConfig {
+		return this.config;
+	}
+
+	public submit(content: any, localOpMetadata: unknown): void {
+		assert(content.type !== upgradeKey, "TODO: reasonable assert here saying incorrect usage");
+		// TODO: Validate content is in correct format.
+		this.pendingOpCount++;
+		this.submitLocalMessage(content, localOpMetadata);
+
+		if (this.configAwaitingSubmission !== undefined) {
```

Comment:
I originally had this submit an upgrade op first to avoid needing to resubmit ops on the client which requested an upgrade, but switched to this because the state necessary to keep around within the collab window is simpler. However, I think it's probably worth the added complexity to use a scheme that doesn't force the client submitting the upgrade to ever write ops in the old format: this allows removing code which writes the old format eventually whereas the current strategy doesn't. There are a few ways we could implement support for sending the upgrade op first--main two approaches would be explicit stamping of the protocol on every op sent, or alternatively make clients in the collab window track any upgrade ops other clients have sent and assume their format is inline with that.